lightning/src/onion_message/messenger.rs

```diff
@@ -1894,98 +1912,72 @@ where
+				let responder = reply_path.map(Responder::new);
+				let response_instructions =
+					self.offers_handler.handle_message(message, context, responder);
+				if let Some((msg, instructions)) = response_instructions {
+					let _ = self.handle_onion_message_response(msg, instructions);
+				}
+			},
+			#[cfg(async_payments)]
+			Ok(PeeledOnion::AsyncPayments(message, context, reply_path)) => {
+				log_receive!(message, reply_path.is_some());
 				let responder = reply_path.map(Responder::new);
 				match message {
-					ParsedOnionMessageContents::Offers(msg) => {
-						let context = match context {
-							None => None,
-							Some(MessageContext::Offers(context)) => Some(context),
-							_ => {
-								debug_assert!(false, "Checked in peel_onion_message");
-								return;
-							},
-						};
-						let response_instructions =
-							self.offers_handler.handle_message(msg, context, responder);
-						if let Some((msg, instructions)) = response_instructions {
-							let _ = self.handle_onion_message_response(msg, instructions);
-						}
-					},
-					#[cfg(async_payments)]
-					ParsedOnionMessageContents::AsyncPayments(
-						AsyncPaymentsMessage::HeldHtlcAvailable(msg),
-					) => {
-						let context = match context {
-							Some(MessageContext::AsyncPayments(context)) => context,
-							Some(_) => {
-								debug_assert!(false, "Checked in peel_onion_message");
-								return;
-							},
-							None => return,
-						};
+					AsyncPaymentsMessage::HeldHtlcAvailable(msg) => {
 						let response_instructions = self
 							.async_payments_handler
 							.handle_held_htlc_available(msg, context, responder);
 						if let Some((msg, instructions)) = response_instructions {
 							let _ = self.handle_onion_message_response(msg, instructions);
 						}
 					},
-					#[cfg(async_payments)]
-					ParsedOnionMessageContents::AsyncPayments(
-						AsyncPaymentsMessage::ReleaseHeldHtlc(msg),
-					) => {
-						let context = match context {
-							Some(MessageContext::AsyncPayments(context)) => context,
-							Some(_) => {
-								debug_assert!(false, "Checked in peel_onion_message");
-								return;
-							},
-							None => return,
-						};
+					AsyncPaymentsMessage::ReleaseHeldHtlc(msg) => {
 						self.async_payments_handler.handle_release_held_htlc(msg, context);
 					},
-					ParsedOnionMessageContents::DNSResolver(DNSResolverMessage::DNSSECQuery(
-						msg,
-					)) => {
+				}
+			},
+			Ok(PeeledOnion::DNSResolver(message, context, reply_path)) => {
+				log_receive!(message, reply_path.is_some());
+				let responder = reply_path.map(Responder::new);
+				match message {
+					DNSResolverMessage::DNSSECQuery(msg) => {
 						let response_instructions =
 							self.dns_resolver_handler.handle_dnssec_query(msg, responder);
 						if let Some((msg, instructions)) = response_instructions {
 							let _ = self.handle_onion_message_response(msg, instructions);
 						}
 					},
```

Comment:
Should we check that there _isn't_ a context for this message?

Reply:
Good q, I somehow was able to dig up an old discussion that covers the same question: https://github.com/lightningdevkit/rust-lightning/pull/3140#discussion_r1676480506 so in that case we went with ignoring the unexpected context since we would've failed to decode if it contained required TLVs. But if you think ignoring the OM is better that's probably fine too.

Reply:
Looks like, with this PR, the case mentioned in that comment now results in ignoring the message. I'd say let's do the same here for consistency.

Reply:
Looks like we'll currently ignore unexpected contexts for `DNSSECQuery` OMs in `main`, so may separate this out into its own commit.

Reply:
That makes sense, pushed a commit for this